### PR TITLE
Source Klaviyo: Remove custom DatetimeBasedCursor dependency

### PR DESCRIPTION
**Pull request by btkcodedev**
## What
Closes https://github.com/airbytehq/airbyte-internal-issues/issues/10604
Remove custom DatetimeBasedCursor dependency

## How
Used stream_interval.start_time as request parameters

### CODE REVIEW

**Comment by natikgadzhi** [comment]
@btkcodedev [some of the tests are unhappy](https://ffa2039b137eac8bafa7224c324c93b48309c6fe982724c4a1f8958-apidata.googleusercontent.com/download/storage/v1/b/airbyte-ci-reports-multi/o/airbyte-ci%2Fconnectors%2Ftest%2Fpull_request%2Fbtkcodedev_fixklaviyocustomcomponents%2F1730930185%2F332c7da1ab8a929faac6a82990dcc03e0588e915%2Fsource-klaviyo%2F2.10.14%2Foutput.html?jk=AXvcXDsomH-yJGW94FMnaLQ481Y4PRmxK7XreDEkZH8JY4QJ6GtgBQjf1yV93dadhj6w2k15C5oYAdlrjEp7eASYFspZXIUj3FHlSlVEOq4vi7f9PkVqNpEqpmZU5c4EGedjqrs--1HJNmF40imjFj-K_y13goTMfxaysvdPp7C6k2WxZdLpla1crcBML9fzc2L6IieOA6lbzru2pVccvoP-0LSyf8Q6KD61MkmXtiTRkMXny5VTSjLeDiCkLBYEyndht4SHiLwqq7M-j0DjLDWRSuGNjlyg4pgVvUyPfIOKrS5aWxOTOqABudKEkKaxVudm9U0AfiNfWKOV9Ps-yIp2nSs5pOJSvgS842XCKz2PyaqRpMRh50XncPrOmENHhAY9SMEveKYanBIaVnHt0DJHtzrDntyoVk4msWdRpWP4n357abkTdEtvx6ISw31nIz55rYOwIVIwwk1Klq5DLVBYD98ZWG2X0tS3ocJ0RE28Yws4zb9grhnlBaPxj2Xw09bp0LpkaDejFGVDpSOoDLB4T1jf8vSDgYEdDDV2sA_69kFhUp-sIP65RngCA1M7MUlEOduL9lG3Zglu7-y5eUBsw1xme13vWqKiIHbniRuwyeM3lW2VrYrIj9KeXLn08EnPBGVu_xGzptH17q0J1A_L7SiTGfhxhb2Zn-sgcYDM1JkdOkJXwz9fX0waslMtI3ZC_hyOl_qH-vQ_kAip7NNPTB_oXgitqWaohSmWsggmvPwXcTfFk9UyBOGC0jXvmj0JTboiwIY57qUV64fh-5iNzCYcIDsicR14S9okVsnVG_xFjW7vx8k95GenA_Vh1GG6kBXHe9fsd9MMly_BfxeQqoiWoP3Kt9IcR7jeM8iDmPRAa07uATr8OWdxmEr0AD7giBGwgFm2ttaQgCc3oQIC_xyifAo41OMz7-7AvJfaPg4cwpiiGbi4OO3peOJ_SbEbIVAboIL8G1cuXx8p-FsEic-3ncWUjYZO6lDwzI4J4Lal5WRYH7Oxgh-wUb8bgSKv4OC_8D0hkhjQ_BRmU0Aae6FTl1uXRudg7wun5lx9aQaLPLos9UkWQXrbS6HUeIKBPglXaTEcgusQvAuRlcx5otwtEUFx3cu3cWjzjBIQ85qpVCoRCicPiposSt76eBhJnCQCetd-I1Lk-tE0TEIHdd4B3rfjJu4DWiRyDw0E6KkPB9tGxuRfnUJGPfjGmZR99G3GtHvZad8U2Uvc_M1cKYHkmxE55v0QnL6D5KLNO7l9yQuWQtAKMW_mVtsR0Vumz2lOaVT3A6sYJjcwmYanjpKMON9X9rIU7N9ViViBHccrS6nsEFSmAa8J5fRwAMMUBYEX149qnGtjMS5smzOQ1SP6pXuKxJgVo3XTMjI8SX25-O7mKmsDlWCXF-Km2PRmFwRLeQ&isca=1) — looks like unit tests are also failing, so that should be the first easy slice to fix.

**Comment by aaronsteers** [comment]
I did a first pass review of the PR and things look more-or-less as expected. That said, I am new to reviewing manifest connectors so I'd appreciate another pair of eyes. As @natikgadzhi points out, there are some failing unit tests and this one specifically looks like the URL params are not in the expected format to match the test scenarios.

`E       requests_mock.exceptions.NoMockAddress: No mock address: GET https://a.klaviyo.com/api/events?fields%5Bmetric%5D=name%2Ccreated%2Cupdated%2Cintegration&include=metric&datetime=2022-11-08T00%3A00%3A00%2B0000`

I am new to this API, but based on the docs, I expect the datetime filter to be wrapped in something like `&filter=greater-or-equal(datetime,'...')`.

https://developers.klaviyo.com/en/reference/get_events

File: airbyte-integrations/connectors/source-klaviyo/source_klaviyo/manifest.yaml
Comment (maxi297) [inline airbyte-integrations/connectors/source-klaviyo/source_klaviyo/manifest.yaml:135]: I haven’t put too much thought into this but it seems like we should have something like:
```
requester:
  request_parameters:
    $ref: <...>
    filter: "greater-than({{ <cursor_field> },{{stream_interval.start_time}})"
    sort: {{ <cursor_field> }}
```

For `<cursor_field>` above, I'm thinking we could have the cursor field as a parameter and interpolate on this. This would avoid duplication (having the cursor_field in the `incremental_sync` component and in the `requester` component. That being said, there might be a better way to avoid this duplication but I'm unsure...

**Comment by btkcodedev** [comment]
@aaronsteers @maxi297 I have committed the changes to manifest,
I have a question. Currently, the main.py file is pointing to the source and streams.py files right?
So the CI tests are based on the python connector and not based on the low code right?

**Comment by maxi297** [comment]
> I have a question. Currently, the main.py file is pointing to the source and streams.py files right?

It seems like there were some streams that could not be ported to low-code [here](https://github.com/airbytehq/airbyte/blob/db95f7a6afd900e86576b453be4f67aa3761f505/airbyte-integrations/connectors/source-klaviyo/source_klaviyo/source.py#L26-L32). This is why there is still a stream.py file. 

> So the CI tests are based on the python connector and not based on the low code right?

There is indeed a `TestCampaignsStream` class that tests the campaigns stream which is not low-code.

**Comment by maxi297** [comment]
@btkcodedev I've fixed some tests. 

### Events tests
Those were failing because we didn't use the proper upper boundary in the manifest.yaml as for the `less-or-equal`, we need to use `stream_interval.end_date`

See https://github.com/airbytehq/airbyte/pull/48391/commits/cf43b0880b3b2910140db5e5ca035205535c83d7

### Profiles tests

This one was a bit annoying. Basically following your change, the `request_options_provider` on the HttpRetriever has been updated to include the incremental parameter. It didn't mean that they weren't sent before but they were added by the layer above i.e. the SimpleRetriever. To avoid this flakiness of "where are the parameters added", I added the parameters as part of the mock server test. The methodology was:
* Checkout master
* Modify airbyte-integrations/connectors/source-klaviyo/unit_tests/integration/test_profiles.py to validate the query params that are being sent
* git stash
* Checkout btkcodedev/fixKlaviyoCustomComponents
* git stash pop
* Run the tests to see if they still pass (and they were passing, yay!)

See https://github.com/airbytehq/airbyte/pull/48391/commits/145758aa633ea1391e38ae80d5b294a3b6a26d3e

@aaronsteers did you want to have a look before I approve?

**Comment by maxi297** [comment]
It seems like the cursor field had changed unexpectedly for events_detailed. I've pushed a change, let's see if CATs pass now! 🤞

File: airbyte-integrations/connectors/source-klaviyo/source_klaviyo/manifest.yaml
Comment (aaronsteers) [inline airbyte-integrations/connectors/source-klaviyo/source_klaviyo/manifest.yaml:296]: @maxi297 - To confirm, should this be greater-than, or greater-or-equal?

I see instances of both in the deleted custom code file.

**Review by aaronsteers — APPROVED** [review]
This looks good and I am happy tests are all passing at this point.

One clarification question inline.

I think the final question is whether we need to run this through regression tests. I can look for a moderate-sized connection to test with.

File: airbyte-integrations/connectors/source-klaviyo/source_klaviyo/manifest.yaml
Comment (maxi297) [reply airbyte-integrations/connectors/source-klaviyo/source_klaviyo/manifest.yaml:296]: My understand is that streams using `KlaviyoCheckpointDatetimeBasedCursor` would have `greater-or-equal` (see https://github.com/airbytehq/airbyte/pull/48391/files#diff-e318ea8c2956167d366ea080f5e4dbb993adfd22cdee41b4b7d1f7dcd36f2d3dL55) while the others would only have `greater_than`. This logic doesn't make sense to me and I would prefer to use `greater-or-equal` (better have duplication than missing records. 

@btkcodedev can we validate if `greater-or-equal` is supported for those streams and if so, update the filters  like @aaronsteers mentioned?

Comment (btkcodedev) [reply airbyte-integrations/connectors/source-klaviyo/source_klaviyo/manifest.yaml:296]: All four changed streams support greater-than and greater-or-equal, for filtering, 
We could wait for this PR to merge and 
Wait for the reflected changes to be successful and make another PR after as it is a certified one.

**Comment by btkcodedev** [comment]
> @btkcodedev I've fixed some tests.

Great work at fixing the tests, Thank you very much @maxi297  :bow:

**Review by maxi297 — APPROVED** [review]
Thanks for addressing the comment!

**Pull request merged**
